File: Threading/C/MatrixMultiplication/matrix_square.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <time.h>

#define SIZE 3

int sMat[SIZE][SIZE];
int rMat[SIZE][SIZE];

typedef struct {
    int row;
} ThreadArg;

void *compute_row(void *arg) {
    int row = ((ThreadArg *)arg)->row;
    for (int col = 0; col < SIZE; col++)
        for (int k = 0; k < SIZE; k++)
            rMat[row][col] += sMat[row][k] * sMat[k][col];
    return NULL;
}

int main(void) {
    srand(time(NULL));

    for (int i = 0; i < SIZE; i++)
        for (int j = 0; j < SIZE; j++)
            sMat[i][j] = rand() % 10;

    printf("Source matrix:\n");
    for (int i = 0; i < SIZE; i++) {
        for (int j = 0; j < SIZE; j++)
            printf("%d ", sMat[i][j]);
        printf("\n");
    }

    pthread_t threads[SIZE];
    ThreadArg args[SIZE];

    for (int i = 0; i < SIZE; i++) {
        args[i].row = i;
        pthread_create(&threads[i], NULL, compute_row, &args[i]);
    }

    for (int i = 0; i < SIZE; i++)
        pthread_join(threads[i], NULL);

    printf("\nResult (squared):\n");
    for (int i = 0; i < SIZE; i++) {
        for (int j = 0; j < SIZE; j++)
            printf("%d ", rMat[i][j]);
        printf("\n");
    }

    return 0;
}

File: Threading/JAVA/MatrixMultiplication/MatrixSquare.java
import java.util.Random;

public class MatrixSquare {

	static int[][] sMat = new int[3][3]; // source matrix
	static int[][] rMat = new int[3][3]; // squared matrix

	public static void main(String [] args)
	{
		Random rand = new Random();

		for (int i = 0; i < sMat.length; i++)
			for (int j = 0; j < sMat[i].length; j++)
				sMat[i][j]=rand.nextInt(10);

		System.out.println("Source matrix is ");
		for (int i = 0; i < sMat.length; i++) 
		{
			for (int j = 0; j < sMat[i].length; j++)
				System.out.print(sMat[i][j]+" ");
			System.out.println();	
		}
		
		MatrixThread thread1 = new MatrixThread(sMat, rMat, 0);
		MatrixThread thread2 = new MatrixThread(sMat, rMat, 1);
		MatrixThread thread3 = new MatrixThread(sMat, rMat, 2);
		
		Thread th1 = new Thread(thread1);
		Thread th2 = new Thread(thread2);
		Thread th3 = new Thread(thread3);		
		try
		{	   
			th1.start();
			th2.start();
			th3.start();

			th1.join();
			th2.join();
			th3.join();
		}
		catch (Exception e) 
		{
			e.printStackTrace();
		}
		System.out.println("\n\nResult:");
		for (int i = 0; i < rMat.length; i++) 
		{
			for (int j = 0; j < rMat[i].length; j++)
				System.out.print(rMat[i][j]+" ");
			System.out.println();
		}
  }
}